### PR TITLE
[Feature] Datacache sql management, only support blacklist in this pr

### PR DESCRIPTION
**Pull request by Smith-Cruise**
## Feature
Support using SQL to manage datacache.
Only support priority=-1 now

## Add rule
Grammar1: `CREATE DATACACHE RULE catalogName.dbName.tblName [WHERE xxxx] PRIORITY=-1;`
Grammar2(select all tables in dbName): `CREATE DATACACHE RULE catalogName.dbName.* PRIORITY=-1;`
Grammar3(select all dbs in catalogName): `CREATE DATACACHE RULE catalogName.*.* PRIORITY=-1;`
Grammar4(select all catalogs): `CREATE DATACACHE RULE *.*.* PRIORITY=-1;`

```sql
mysql> create datacache rule hive_test.chendingchao.datacache_par_string where dt>'2023-09-13' priority=-1;
Query OK, 0 rows affected (0.45 sec)
```

## Show rules
```sql
mysql> show datacache rules;
+---------+-----------+--------------+----------------------+----------+-----------------------------------------------------------------------+------------+
| Rule Id | Catalog   | Database     | Table                | Priority | Predicates                                                            | Properties |
+---------+-----------+--------------+----------------------+----------+-----------------------------------------------------------------------+------------+
| 0       | hive_test | chendingchao | datacache_par_string | -1       | `hive_test`.`chendingchao`.`datacache_par_string`.`dt` > '2023-09-13' | NULL       |
+---------+-----------+--------------+----------------------+----------+-----------------------------------------------------------------------+------------+
1 row in set (0.01 sec)
```

## Drop rule
```sql
drop datacache rule 0;
```

## Clear all rules
```sql
clear datacache rules;
```

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [x] Yes, this PR will result in a change in behavior.
- [ ] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [x] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [x] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 203 / 310 (65.48%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|4|00.00%|[742, 746, 750, 754]|
|:large_blue_circle:|com/starrocks/sql/ast/DropDataCacheRuleStmt.java|0|5|00.00%|[24, 25, 26, 29, 34]|
|:large_blue_circle:|com/starrocks/sql/ast/ClearDataCacheRulesStmt.java|0|3|00.00%|[22, 23, 27]|
|:large_blue_circle:|com/starrocks/sql/ast/ShowDataCacheRulesStmt.java|0|14|00.00%|[24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 36, 37, 41, 46]|
|:large_blue_circle:|com/starrocks/qe/DDLStmtExecutor.java|0|13|00.00%|[967, 968, 969, 970, 971, 976, 977, 978, 979, 984, 985, 986, 987]|
|:large_blue_circle:|com/starrocks/qe/ShowExecutor.java|1|5|20.00%|[386, 2586, 2587, 2588]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzerVisitor.java|2|8|25.00%|[721, 722, 727, 728, 733, 734]|
|:large_blue_circle:|com/starrocks/datacache/DataCacheRule.java|13|36|36.11%|[40, 56, 60, 61, 62, 63, 64, 65, 66, 67, 69, 71, 72, 74, 76, 80, 81, 82, 83, 84, 85, 87, 89]|
|:large_blue_circle:|com/starrocks/datacache/DataCacheMgr.java|66|83|79.52%|[66, 75, 90, 111, 121, 141, 151, 167, 168, 169, 170, 173, 175, 176, 177, 178, 179]|
|:large_blue_circle:|com/starrocks/connector/RemoteScanRangeLocations.java|36|45|80.00%|[88, 184, 185, 186, 187, 297, 298, 299, 308]|
|:large_blue_circle:|com/starrocks/datacache/DataCacheOptions.java|5|6|83.33%|[27]|
|:large_blue_circle:|com/starrocks/sql/analyzer/DataCacheStmtAnalyzer.java|54|61|88.52%|[86, 104, 105, 106, 108, 113, 148]|
|:large_blue_circle:|com/starrocks/datacache/DataCacheExprRewriter.java|13|14|92.86%|[58]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/ScalarOperator.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/CreateDataCacheRuleStmt.java|11|11|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 1 / 3 (33.33%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/connector/hive_connector.cpp|1|3|33.33%|[102, 103]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 270 / 312 (86.54%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|4|00.00%|[742, 746, 750, 754]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzerVisitor.java|2|8|25.00%|[721, 722, 727, 728, 733, 734]|
|:large_blue_circle:|com/starrocks/qe/DDLStmtExecutor.java|8|13|61.54%|[967, 968, 969, 970, 971]|
|:large_blue_circle:|com/starrocks/connector/RemoteScanRangeLocations.java|37|48|77.08%|[88, 184, 185, 186, 187, 239, 240, 303, 304, 305, 314]|
|:large_blue_circle:|com/starrocks/datacache/DataCacheOptions.java|5|6|83.33%|[27]|
|:large_blue_circle:|com/starrocks/sql/analyzer/DataCacheStmtAnalyzer.java|54|60|90.00%|[104, 105, 106, 108, 113, 150]|
|:large_blue_circle:|com/starrocks/datacache/DataCacheMgr.java|77|83|92.77%|[69, 78, 93, 114, 144, 154]|
|:large_blue_circle:|com/starrocks/datacache/DataCacheExprRewriter.java|13|14|92.86%|[59]|
|:large_blue_circle:|com/starrocks/sql/ast/ShowDataCacheRulesStmt.java|13|14|92.86%|[41]|
|:large_blue_circle:|com/starrocks/datacache/DataCacheRule.java|35|36|97.22%|[56]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/ScalarOperator.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/DropDataCacheRuleStmt.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/CreateDataCacheRuleStmt.java|11|11|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/ClearDataCacheRulesStmt.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/ShowExecutor.java|5|5|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 1 / 3 (33.33%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/connector/hive_connector.cpp|1|3|33.33%|[102, 103]|

File: be/src/connector/hive_connector.cpp [hunk 1]
```diff
@@ -97,6 +97,11 @@ Status HiveDataSource::open(RuntimeState* state) {
     if (state->query_options().__isset.enable_populate_block_cache) {
         _enable_populate_block_cache = state->query_options().enable_populate_block_cache;
     }
+    // Close block cache when priority = -1
+    if (_scan_range.__isset.data_cache_options && _scan_range.data_cache_options.__isset.priority &&
```
[the remark this answers is not in the document]
Comment (Smith-Cruise) [reply be/src/connector/hive_connector.cpp:101]: good catch

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/AstBuilder.java [hunk 2]
```diff
@@ -2928,6 +2932,53 @@ public ParseNode visitShowWhiteListStatement(StarRocksParser.ShowWhiteListStatem
...
 
+    // --------------------------------------- Data Cache Management Statement -----------------------------------------
+    @Override
+    public ParseNode visitCreateDataCacheRuleStatement(StarRocksParser.CreateDataCacheRuleStatementContext ctx) {
+        List<StarRocksParser.IdentifierOrStringOrStarContext> partList = ctx.dataCacheTarget().identifierOrStringOrStar();
+        List<String> parts = partList.stream().map(c -> ((Identifier) visit(c)).getValue()).collect(toList());
+
+        QualifiedName qualifiedName = QualifiedName.of(parts);
+
+        int priority = Integer.parseInt(ctx.INTEGER_VALUE().getText());
```
[the remark this answers is not in the document]
Comment (Smith-Cruise) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/AstBuilder.java:2943]: yes, `INTEGER_VALUE()` always return `>=0`

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/DataCacheStmtAnalyzer.java [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    }
+
+    private static Optional<Table> getTable(String catalogName, String dbName, String tblName) {
+        MetadataMgr metadataMgr = GlobalStateMgr.getCurrentState().getMetadataMgr();
+
+        // check validity
+        if (isSelectAll(catalogName)) {
+            if (!isSelectAll(dbName) || !isSelectAll(tblName)) {
+                throw new SemanticException("Catalog is *, database and table must use * either");
+            }
```
[the remark this answers is not in the document]
Comment (Smith-Cruise) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/DataCacheStmtAnalyzer.java:128]: We need to check in db level further.

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/DataCacheStmtAnalyzer.java [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+            return null;
+        }
+
+        @Override
+        public Void visitDropDataCacheRuleStatement(DropDataCacheRuleStmt statement, ConnectContext context) {
+            long cacheRuleId = statement.getCacheRuleId();
+            if (!dataCacheMgr.isExistCacheRule(cacheRuleId)) {
+                throw new SemanticException(String.format("Datacache rule id = %d not existed", cacheRuleId));
```
[the remark this answers is not in the document]
Comment (Smith-Cruise) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/DataCacheStmtAnalyzer.java:106]: changed

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/DataCacheStmtAnalyzer.java [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+            // check target existed
+            String catalogName = parts.get(0);
+            String dbName = parts.get(1);
+            String tblName = parts.get(2);
+
+            if (CatalogMgr.isInternalCatalog(catalogName)) {
+                throw new SemanticException("DataCache only support external catalog now");
+            }
+
+            Optional<Table> optionalTable = getTable(catalogName, dbName, tblName);
```
[the remark this answers is not in the document]
Comment (Smith-Cruise) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/DataCacheStmtAnalyzer.java:77]: getTable may throw exception, if people specific none-existed table, `getTable()` will throw exception.

File: fe/fe-core/src/main/java/com/starrocks/datacache/DataCacheMgr.java [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+        return idToCacheRuleMap.containsKey(id);
+    }
+
+    public synchronized Optional<DataCacheRule> getCacheRule(String catalogName, String dbName, String tblName) {
+        // check catalog level
+        ConcurrentMap<String, ConcurrentMap<String, DataCacheRule>> catalogMap;
+        if (nameToCacheRuleMap.containsKey(STAR_MATCH_ALL)) {
+            catalogMap = nameToCacheRuleMap.get(STAR_MATCH_ALL);
+        } else if (nameToCacheRuleMap.containsKey(catalogName)) {
+            catalogMap = nameToCacheRuleMap.get(catalogName);
```
[the remark this answers is not in the document]
Comment (Smith-Cruise) [reply fe/fe-core/src/main/java/com/starrocks/datacache/DataCacheMgr.java]: changed

File: fe/fe-core/src/main/java/com/starrocks/datacache/DataCacheMgr.java [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+        } else {
+            dbMap = catalogMap.get(dbName);
+        }
+
+        dbMap.put(tblName, cacheRule);
+    }
+
+    public synchronized void dropCacheRule(long id) throws SemanticException {
+        DataCacheRule cacheRule = idToCacheRuleMap.remove(id);
+        String catalogName = cacheRule.getTarget().getParts().get(0);
```
[the remark this answers is not in the document]
Comment (Smith-Cruise) [reply fe/fe-core/src/main/java/com/starrocks/datacache/DataCacheMgr.java]: changed

File: fe/fe-core/src/main/java/com/starrocks/datacache/DataCacheMgr.java [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+            catalogMap = nameToCacheRuleMap.get(catalogName);
+        }
+
+        // Put in db level
+        ConcurrentMap<String, DataCacheRule> dbMap;
+        if (!catalogMap.containsKey(dbName)) {
+            dbMap = new ConcurrentHashMap<String, DataCacheRule>();
+            catalogMap.put(dbName, dbMap);
+        } else {
+            dbMap = catalogMap.get(dbName);
```
[the remark this answers is not in the document]
Comment (Smith-Cruise) [reply fe/fe-core/src/main/java/com/starrocks/datacache/DataCacheMgr.java]: changed

File: fe/fe-core/src/main/java/com/starrocks/datacache/DataCacheMgr.java [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+        String dbName = target.getParts().get(1);
+        String tblName = target.getParts().get(2);
+
+        // Put in catalog level
+        ConcurrentMap<String, ConcurrentMap<String, DataCacheRule>> catalogMap;
+        if (!nameToCacheRuleMap.containsKey(catalogName)) {
+            catalogMap = new ConcurrentHashMap<>();
+            nameToCacheRuleMap.put(catalogName, catalogMap);
+        } else {
+            catalogMap = nameToCacheRuleMap.get(catalogName);
```
[the remark this answers is not in the document]
Comment (Smith-Cruise) [reply fe/fe-core/src/main/java/com/starrocks/datacache/DataCacheMgr.java]: changed

File: fe/fe-core/src/main/java/com/starrocks/datacache/DataCacheMgr.java [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+public class DataCacheMgr {
+
+    private static final DataCacheMgr INSTANCE = new DataCacheMgr();
+
+    private final AtomicLong ids = new AtomicLong();
+    private final ConcurrentMap<Long, DataCacheRule> idToCacheRuleMap = new ConcurrentHashMap<>();
+    // key: catalog name => value: ConcurrentMap<String, ConcurrentMap<String, DataCacheRule>>
+    // key: db name =>  value: ConcurrentMap<String, DataCacheRule>
+    // key: table name => value: DataCacheRule
+    private final ConcurrentMap<String, ConcurrentMap<String, ConcurrentMap<String, DataCacheRule>>>
```
[the remark this answers is not in the document]
Comment (Smith-Cruise) [reply fe/fe-core/src/main/java/com/starrocks/datacache/DataCacheMgr.java]: changed

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 1 / 3 (33.33%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/connector/hive_connector.cpp|1|3|33.33%|[102, 103]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 1 / 3 (33.33%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/connector/hive_connector.cpp|1|3|33.33%|[102, 103]|

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/DataCacheStmtAnalyzer.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (GavinMar) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/DataCacheStmtAnalyzer.java:132]: > We need to check in db level further.

If all parts are "*", why still need check in db level ?

File: fe/fe-core/src/main/java/com/starrocks/datacache/DataCacheMgr.java [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class DataCacheMgr {
+
+    private static final DataCacheMgr INSTANCE = new DataCacheMgr();
+
+    private final AtomicLong ids = new AtomicLong();
+    private final ConcurrentMap<Long, DataCacheRule> idToCacheRuleMap = new ConcurrentHashMap<>();
```
[the remark this answers is not in the document]
Comment (Smith-Cruise) [reply fe/fe-core/src/main/java/com/starrocks/datacache/DataCacheMgr.java]: changed

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/DataCacheStmtAnalyzer.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Smith-Cruise) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/DataCacheStmtAnalyzer.java:132]: fixed

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 310 / 352 (88.07%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|4|00.00%|[738, 742, 746, 750]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzerVisitor.java|2|8|25.00%|[721, 722, 727, 728, 733, 734]|
|:large_blue_circle:|com/starrocks/qe/DDLStmtExecutor.java|8|13|61.54%|[972, 973, 974, 975, 976]|
|:large_blue_circle:|com/starrocks/connector/RemoteScanRangeLocations.java|37|48|77.08%|[88, 184, 185, 186, 187, 239, 240, 303, 304, 305, 314]|
|:large_blue_circle:|com/starrocks/datacache/DataCacheOptions.java|5|6|83.33%|[27]|
|:large_blue_circle:|com/starrocks/sql/analyzer/DataCacheStmtAnalyzer.java|57|63|90.48%|[107, 108, 109, 111, 116, 159]|
|:large_blue_circle:|com/starrocks/sql/ast/ShowDatacacheRulesStmt.java|13|14|92.86%|[41]|
|:large_blue_circle:|com/starrocks/datacache/DataCacheExprRewriter.java|13|14|92.86%|[59]|
|:large_blue_circle:|com/starrocks/datacache/DataCacheMgr.java|114|120|95.00%|[105, 108, 110, 133, 165, 175]|
|:large_blue_circle:|com/starrocks/datacache/DataCacheRule.java|35|36|97.22%|[56]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/ScalarOperator.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/ClearDatacacheRulesStmt.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/CreateDatacacheRuleStmt.java|11|11|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/ShowExecutor.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/DropDatacacheRuleStmt.java|5|5|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 1 / 3 (33.33%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/connector/hive_connector.cpp|1|3|33.33%|[102, 103]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 310 / 352 (88.07%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|4|00.00%|[738, 742, 746, 750]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzerVisitor.java|2|8|25.00%|[721, 722, 727, 728, 733, 734]|
|:large_blue_circle:|com/starrocks/qe/DDLStmtExecutor.java|8|13|61.54%|[972, 973, 974, 975, 976]|
|:large_blue_circle:|com/starrocks/connector/RemoteScanRangeLocations.java|37|48|77.08%|[88, 184, 185, 186, 187, 239, 240, 303, 304, 305, 314]|
|:large_blue_circle:|com/starrocks/datacache/DataCacheOptions.java|5|6|83.33%|[27]|
|:large_blue_circle:|com/starrocks/sql/analyzer/DataCacheStmtAnalyzer.java|57|63|90.48%|[107, 108, 109, 111, 116, 159]|
|:large_blue_circle:|com/starrocks/sql/ast/ShowDatacacheRulesStmt.java|13|14|92.86%|[41]|
|:large_blue_circle:|com/starrocks/datacache/DataCacheExprRewriter.java|13|14|92.86%|[59]|
|:large_blue_circle:|com/starrocks/datacache/DataCacheMgr.java|114|120|95.00%|[105, 108, 110, 133, 165, 175]|
|:large_blue_circle:|com/starrocks/datacache/DataCacheRule.java|35|36|97.22%|[56]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/ScalarOperator.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/ClearDatacacheRulesStmt.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/CreateDatacacheRuleStmt.java|11|11|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/ShowExecutor.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/DropDatacacheRuleStmt.java|5|5|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 1 / 3 (33.33%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/connector/hive_connector.cpp|1|3|33.33%|[102, 103]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 1 / 3 (33.33%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/connector/hive_connector.cpp|1|3|33.33%|[102, 103]|

File: be/src/connector/hive_connector.cpp [hunk 12]
```diff
@@ -97,6 +97,11 @@ Status HiveDataSource::open(RuntimeState* state) {
     if (state->query_options().__isset.enable_populate_block_cache) {
         _enable_populate_block_cache = state->query_options().enable_populate_block_cache;
     }
+    // Close block cache when priority = -1
```
Comment (letian-jiang) [inline be/src/connector/hive_connector.cpp:100]: `Close block cache` is confusing

**Review by letian-jiang — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 317 / 359 (88.30%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|4|00.00%|[738, 742, 746, 750]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzerVisitor.java|2|8|25.00%|[721, 722, 727, 728, 733, 734]|
|:large_blue_circle:|com/starrocks/qe/DDLStmtExecutor.java|8|13|61.54%|[972, 973, 974, 975, 976]|
|:large_blue_circle:|com/starrocks/connector/RemoteScanRangeLocations.java|39|50|78.00%|[88, 184, 185, 186, 187, 242, 243, 306, 307, 308, 317]|
|:large_blue_circle:|com/starrocks/datacache/DatacacheOptions.java|5|6|83.33%|[27]|
|:large_blue_circle:|com/starrocks/sql/analyzer/DatacacheStmtAnalyzer.java|57|63|90.48%|[107, 108, 109, 111, 116, 159]|
|:large_blue_circle:|com/starrocks/sql/ast/ShowDatacacheRulesStmt.java|13|14|92.86%|[41]|
|:large_blue_circle:|com/starrocks/datacache/DatacacheExprRewriter.java|13|14|92.86%|[59]|
|:large_blue_circle:|com/starrocks/datacache/DatacacheMgr.java|114|120|95.00%|[105, 108, 110, 133, 165, 175]|
|:large_blue_circle:|com/starrocks/datacache/DatacacheRule.java|35|36|97.22%|[56]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/ScalarOperator.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/ClearDatacacheRulesStmt.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/CreateDatacacheRuleStmt.java|11|11|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/ShowExecutor.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/DropDatacacheRuleStmt.java|5|5|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 1 / 3 (33.33%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/connector/hive_connector.cpp|1|3|33.33%|[102, 103]|

File: fe/fe-core/src/main/java/com/starrocks/connector/RemoteScanRangeLocations.java [hunk 13]
```diff
@@ -14,19 +14,31 @@
...
 import com.google.common.collect.Lists;
 import com.starrocks.analysis.DescriptorTable;
+import com.starrocks.analysis.Expr;
 import com.starrocks.catalog.HiveMetaStoreTable;
 import com.starrocks.catalog.HiveTable;
 import com.starrocks.catalog.HudiTable;
 import com.starrocks.catalog.PartitionKey;
 import com.starrocks.catalog.Table;
 import com.starrocks.common.Config;
+import com.starrocks.datacache.DatacacheExprRewriter;
```
Comment (GavinMar) [inline fe/fe-core/src/main/java/com/starrocks/connector/RemoteScanRangeLocations.java:28]: You'd better use the `DataCache` prefix to keep same name format with lake module.

**Review by GavinMar — APPROVED** [review]

**Review by wangruin — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 [hunk 14]
```diff
@@ -2627,7 +2655,7 @@ nonReserved
     | ARRAY_AGG
     | BACKEND | BACKENDS | BACKUP | BEGIN | BITMAP_UNION | BLACKLIST | BINARY | BODY | BOOLEAN | BROKER | BUCKETS
     | BUILTIN | BASE
-    | CAST | CANCEL | CATALOG | CATALOGS | CEIL | CHAIN | CHARSET | CLEAN | CLUSTER | CLUSTERS | CURRENT | COLLATION | COLUMNS
+    | CAST | CANCEL | CATALOG | CATALOGS | CEIL | CHAIN | CHARSET | CLEAN | CLEAR | CLUSTER | CLUSTERS | CURRENT | COLLATION | COLUMNS
     | CUME_DIST | CUMULATIVE | COMMENT | COMMIT | COMMITTED | COMPUTE | CONNECTION | CONSISTENT | COSTS | COUNT
     | CONFIG | COMPACT
     | DATA | DATE | DATETIME | DAY | DECOMMISSION | DISABLE | DISTRIBUTION | DUPLICATE | DYNAMIC | DISTRIBUTED | DEALLOCATE
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:2661]: add DATACACHE

File: fe/fe-core/src/main/java/com/starrocks/connector/RemoteScanRangeLocations.java (hunk 13 above)
Comment (Smith-Cruise) [reply fe/fe-core/src/main/java/com/starrocks/connector/RemoteScanRangeLocations.java:28]: changed in next pr

File: be/src/connector/hive_connector.cpp (hunk 12 above)
Comment (Smith-Cruise) [reply be/src/connector/hive_connector.cpp]: changed

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 (hunk 14 above)
Comment (Smith-Cruise) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4]: added

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 317 / 359 (88.30%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|4|00.00%|[738, 742, 746, 750]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzerVisitor.java|2|8|25.00%|[721, 722, 727, 728, 733, 734]|
|:large_blue_circle:|com/starrocks/qe/DDLStmtExecutor.java|8|13|61.54%|[972, 973, 974, 975, 976]|
|:large_blue_circle:|com/starrocks/connector/RemoteScanRangeLocations.java|39|50|78.00%|[88, 184, 185, 186, 187, 242, 243, 306, 307, 308, 317]|
|:large_blue_circle:|com/starrocks/datacache/DatacacheOptions.java|5|6|83.33%|[27]|
|:large_blue_circle:|com/starrocks/sql/analyzer/DatacacheStmtAnalyzer.java|57|63|90.48%|[107, 108, 109, 111, 116, 159]|
|:large_blue_circle:|com/starrocks/sql/ast/ShowDatacacheRulesStmt.java|13|14|92.86%|[41]|
|:large_blue_circle:|com/starrocks/datacache/DatacacheExprRewriter.java|13|14|92.86%|[59]|
|:large_blue_circle:|com/starrocks/datacache/DatacacheMgr.java|114|120|95.00%|[105, 108, 110, 133, 165, 175]|
|:large_blue_circle:|com/starrocks/datacache/DatacacheRule.java|35|36|97.22%|[56]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/ScalarOperator.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/ClearDatacacheRulesStmt.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/CreateDatacacheRuleStmt.java|11|11|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/ShowExecutor.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/DropDatacacheRuleStmt.java|5|5|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 1 / 3 (33.33%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/connector/hive_connector.cpp|1|3|33.33%|[102, 103]|

**Review by packy92 — APPROVED** [review]

**Review by dirtysalt — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/connector/RemoteScanRangeLocations.java (hunk 13 above)
Comment (Smith-Cruise) [reply fe/fe-core/src/main/java/com/starrocks/connector/RemoteScanRangeLocations.java:28]: changed in #32222

**Pull request merged**